lib/AST/NameLookup.cpp

```diff
@@ -361,6 +361,50 @@ enum class ConstructorComparison {
   Better,
 };
 
+bool swift::removeOutOfModuleDecls(SmallVectorImpl<ValueDecl*> &decls,
```

Comment:
It seems that it would be better to implement this so that the out-of-module decls do not appear in the result list in the first place?

Reply:
As for the meat of your question:

* When doing qualified lookups, filtering is basically the only viable implementation since `MemberLookupTable` is flat and isn't keyed on module. (Adding module information to it would require many duplicate entries; having a member lookup table per type/module would slow down non-module-selector lookups.)
* When doing unqualified lookups, `UnqualifiedLookupFactory::addImportedResults()` attempts to restrict the search to just the modules matched by the module selector. Filtering is *probably* redundant there, but it may catch edge cases where candidates get added to lookup through weird paths.